src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -231,21 +237,270 @@ def _estimate_inplane_rotations(
     # Secondary Methods for computing outer product #
     #################################################
 
-    def _self_clmatrix_c3_c4(self, pf, order, max_shift, shift_step):
-        # return sclmatrix
-        pass
+    def _self_clmatrix_c3_c4(self):
+        """
+        Find the single pair of self-common-lines in each image assuming that the underlying
+        symmetry is C3 or C4.
+        """
+        pf = self.pf.copy()
+        n_img = self.n_img
+        n_theta = self.n_theta
+        max_shift_1d = np.ceil(2 * np.sqrt(2) * self.max_shift)
+        shift_step = self.shift_step
+        order = self.order
 
-    def _estimate_all_Riis_c3_c4(order, sclmatrix, n_theta):
-        # return Riis
-        pass
+        # The angle between self-common-lines is in the range [60, 180] for C3 symmetry
+        # and [90, 180] for C4 symmetry. Since antipodal lines are perfectly correlated
+        # we search for common lines in a smaller window.
+        # Note: matlab code used [60, 165] for order=3 and [90, 160] for order=4.
+        # We've increased this window to improve the MSE of our estimates.
+        if order == 3:
+            min_angle_diff = 60 * np.pi / 180
+            max_angle_diff = 175 * np.pi / 180
+        else:
+            min_angle_diff = 90 * np.pi / 180
+            max_angle_diff = 175 * np.pi / 180
+
+        # The self-common-lines matrix holds two indices per image that represent
+        # the two self common-lines in the image.
+        sclmatrix = np.zeros((n_img, 2))
+        corrs_stats = np.zeros(n_img)
```

Comment:
These will store the correlation and shift values. They were used for diagnostics in the matlab code. I left them there in case they would be useful for testing.

Reply:
Unless there's an immediate use case, I would remove them. Each time we call the function, we discard these outputs. If someone wants to debug, they can put these things in rather easily.

Reply:
Sounds good. Took them out.